src/librustc_lint/types.rs

```diff
@@ -152,11 +152,52 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TypeLimits {
 
                                 // Detect literal value out of range [min, max] inclusive
                                 // avoiding use of -min to prevent overflow/panic
-                                if (negative && v > max + 1) ||
-                                   (!negative && v > max) {
-                                    cx.span_lint(OVERFLOWING_LITERALS,
-                                                 e.span,
-                                                 &format!("literal out of range for {:?}", t));
+                                if (negative && v > max + 1) || (!negative && v > max) {
+                                    if let Some(repr_str) = get_bin_hex_repr(cx, lit) {
+                                        let bits = int_ty_bits(t, cx.sess().target.isize_ty);
+                                        let mut actually = v as i128;
+                                        if bits < 128 {
+                                            // v & 0b0..01..1, |1| = bits
+                                            let trimmed = v & ((1 << bits) - 1);
+                                            actually = if v & (1 << (bits - 1)) == 0 {
```

Comment:
You can simplify these conversions by first casting to `i128` and then bitshifting by `128 - bits` to the left and then to the right.